rfcs/027-pipeline-intermediate-storage/README.md

```diff
@@ -0,0 +1,147 @@
+
+* No transactions which might prove useful for certain use cases.
+
+Questions:
+
+* Elasticsearch supports [versioning](https://www.elastic.co/blog/elasticsearch-versioning-support), including the ability to index with our own external version numbers. However when using external versioning it only indexes data when the version number is greater than currently stored, whereas our business logic currently expects updates to be processed when versions are greater or equal. Some thought will need to be given whether this might work for us, and if not we would need to implement our own system on top of the Elasticsearch APIs (which may not be particularly resilient to race conditions due to lack of transactions).
+
+* Eventual consistency is probably sufficient for our purposes, but may need some investigation. The [refresh interval](#https://www.elastic.co/guide/en/elasticsearch/reference/current/tune-for-indexing-speed.html#_unset_or_increase_the_refresh_interval) where written data is made available is 1 second by default, which should be enough that downstream services see the latest data, but it might be worth being more aggressive than this.
+
+### Redis
+
+Pros:
+
+* In memory so likely to be fastest reads / writes out of all the options.
+
+* AWS offers it via ElastiCache.
+
+* Improved visibility over VHS.
+
+Cons:
+
+* Requires an instance with enough memory to store all the data, so whilst certainly possible with our current sizes of data the point at which it becomes infeasible will be reached much quicker than other solutions.
+
+* No inbuilt versioning so would have to implement something manually within our applications.
+
+* Visibility nothing like Kibana, and not particularly queryable.
+
+Questions:
+
+* Redis offers some persistence, either via regular snapshots or by logging individual create / update / delete operations to disk with the ability to reconstruct the store via playback. Would need further investigation.
+
+* Pricing is unknown, likely would need some pretty large reserved EC2 instances but expect it to be cheaper than DynamoDB / S3 combination.
+
+### AWS RDS / PostgreSQL
+
+Pros:
+
+* The most reliable and predictable option due to being the only solution which is completely ACID compliant.
+
+* Can support our versioning needs (this could be achieved e.g. within our application logic using transactions, or within a more complex single query using clauses).
+
+* Available as a service on AWS.
+
+* Improved visibility over VHS, and ability to query data.
+
+* Replicas can help with scalability when large number of reads.
+
+* Transaction support might be useful for some of our current logic or future use cases.
+
+Cons:
+
+* Performance may not match other options due to stronger consistency guarantees (though likely to be much better than VHS due to only hitting a single system and making use of a cache).
```

Comment:
I expect given ID minter experience that there might also be throughput issues (ie all the general problems people have with scaling a RDBMS)

Reply:
I know there is no desire to use Postgres / RDS, and I don't personally think we should for this purpose, but just to stick up for them: I don't know about the ID minter in particular but in the postgres through bulk inserts you can write 100K+ rows per second, and have no doubt we could make it much faster that DynamoDB + S3 (where each item is written individually with a call to two separate services)